modules/ballerina-core/src/main/java/org/ballerinalang/model/builder/BLangModelBuilder.java

```diff
@@ -2166,7 +2166,24 @@ private void validateIdentifier(String identifier, NodeLocation location) {
     private void validateTransformStatementBody(BlockStmt blockStmt, Map<String, Expression> inputs,
                                                 Map<String, Expression> outputs) {
         for (Statement statement : blockStmt.getStatements()) {
-            if (statement instanceof AssignStmt) {
+            if (statement instanceof VariableDefStmt) {
+                VariableDefStmt variableDefStmt = (VariableDefStmt) statement;
+                Expression varRefExpression = variableDefStmt.getLExpr();
+                String varName = ((SimpleVarRefExpr) varRefExpression).getVarName();
+                //variables defined in transform scope, cannot be used as output
+                if (outputs.get(varName) == null) {
+                    //if variable has not been used as an output before
+                    if (inputs.get(varName) == null) {
```

Comment:
Shall we refactor this to something like 
if(abc == null) {
   return;
}
not null logic...
if(cde == null) {
   return;
}
not null logic